### PR TITLE
Serialize content of style information element with URL replacement

### PR DESCRIPTION
**Pull request by szewai**
#### 3520ab7bac786d2b9d2bac921f67b0dfdddda6b7
<pre>
Serialize content of style information element with URL replacement
<a href="https://bugs.webkit.org/show_bug.cgi?id=264022">https://bugs.webkit.org/show_bug.cgi?id=264022</a>
<a href="https://rdar.apple.com/116883064">rdar://116883064</a>

Reviewed by Ryosuke Niwa.

Instead of directly appending plain text content of style information element to markup, building up a contents string
from rules in the element and appending the string to markup. In this way, we could replace the subresource URLs on
demand.

API test: WebArchive.SaveResourcesStyle

* Source/WebCore/css/CSSFontFaceRule.cpp:
(WebCore::CSSFontFaceRule::cssText const):
(WebCore::CSSFontFaceRule::cssTextWithReplacementURLs const):
(WebCore::CSSFontFaceRule::cssTextInternal const):
* Source/WebCore/css/CSSFontFaceRule.h:
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
(WebCore::CSSFontFaceSrcResourceValue::customSetReplacementURLForSubresources):
(WebCore::CSSFontFaceSrcResourceValue::customClearReplacementURLForSubresources):
(WebCore::CSSFontFaceSrcResourceValue::customCSSText const):
* Source/WebCore/css/CSSFontFaceSrcValue.h:
* Source/WebCore/css/CSSImageValue.cpp:
(WebCore::CSSImageValue::customSetReplacementURLForSubresources):
(WebCore::CSSImageValue::customClearReplacementURLForSubresources):
(WebCore::CSSImageValue::customCSSText const):
* Source/WebCore/css/CSSImageValue.h:
* Source/WebCore/css/CSSRule.h:
(WebCore::CSSRule::cssTextWithReplacementURLs const):
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::cssText const):
(WebCore::CSSStyleRule::cssTextForRules const):
(WebCore::CSSStyleRule::cssTextWithReplacementURLs const):
(WebCore::CSSStyleRule::cssTextForRulesWithReplacementURLs const):
(WebCore::CSSStyleRule::cssTextInternal const):
(WebCore::CSSStyleRule::cssTextForDeclsAndRules const): Deleted.
* Source/WebCore/css/CSSStyleRule.h:
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::setReplacementURLForSubresources):
(WebCore::CSSValue::clearReplacementURLForSubresources):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::customSetReplacementURLForSubresources):
(WebCore::CSSValue::customClearReplacementURLForSubresources):
* Source/WebCore/css/CSSValueList.cpp:
(WebCore::CSSValueContainingVector::customSetReplacementURLForSubresources):
(WebCore::CSSValueContainingVector::customClearReplacementURLForSubresources):
* Source/WebCore/css/CSSValueList.h:
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::setReplacementURLForSubresources):
(WebCore::StyleProperties::clearReplacementURLForSubresources):
* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::appendContentsForNode):
(WebCore::MarkupAccumulator::serializeNodesWithNamespaces):
* Source/WebCore/editing/MarkupAccumulator.h:
* Source/WebCore/html/HTMLStyleElement.cpp:
(WebCore::HTMLStyleElement::addSubresourceAttributeURLs const):
(WebCore::HTMLStyleElement::textContentWithReplacementURLs const):
* Source/WebCore/html/HTMLStyleElement.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: <a href="https://commits.webkit.org/270114@main">https://commits.webkit.org/270114@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/97e77bb3ac9871f693f557f3a087deabf71cee2e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24594 "3 style errors") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3143 "Hash 97e77bb3 for PR 19819 does not build (failure)") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25851 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26717 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22593 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24863 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4811 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/577 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22979 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24839 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2206 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21237 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27300 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1929 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22166 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28346 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22439 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22501 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26154 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1848 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/170 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3156 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5896 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2303 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2218 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/editing/MarkupAccumulator.cpp [hunk 1]
```diff
@@ -208,6 +209,17 @@ String MarkupAccumulator::serializeNodes(Node& targetNode, SerializedNodes root,
     return m_markup.toString();
 }
 
+bool MarkupAccumulator::appendContentsForNode(StringBuilder& result, const Node& targetNode)
+{
+    if (m_replacementURLStrings.isEmpty() || !targetNode.hasTagName(styleTag))
+        return false;
+
+    auto& styleElement = downcast<HTMLStyleElement>(targetNode);
+    result.append(styleElement.contentsStringWithReplacementURLs(m_replacementURLStrings));
```
Comment (rniwa) [inline Source/WebCore/editing/MarkupAccumulator.cpp]: I'd call this `textContentWithReplacementURLs` or `stringWithReplacementURLs` instead.

File: Source/WebCore/editing/MarkupAccumulator.cpp [hunk 2]
```diff
@@ -248,8 +260,9 @@ void MarkupAccumulator::serializeNodesWithNamespaces(Node& targetNode, Serialize
                 }
             }
 
+            bool shouldSkipChidren = appendContentsForNode(m_markup, *current);
```
Comment (rniwa) [inline Source/WebCore/editing/MarkupAccumulator.cpp:262]: It's not great that we're making the traversal logic more complicated but I don't have a better suggestion.

**Comment by rniwa** [comment]
So the term "inline style" is reserved for `style` content attribute you can specify on any element so it's rather confusing to call this "inline URLs inside style information element". Maybe rephrase it like "Serialize contents of style elements with URL replacements" or something.

**Review by rniwa — APPROVED** [review]

**Comment by rniwa** [comment]
I'm assuming you'd be fixing `link` element and `style` content attribute in a follow up?

**Comment by szewai** [comment]
> So the term "inline style" is reserved for `style` content attribute you can specify on any element so it's rather confusing to call this "inline URLs inside style information element". Maybe rephrase it like "Serialize contents of style elements with URL replacements" or something.

Sure, I will update the title.

**Comment by szewai** [comment]
> I'm assuming you'd be fixing `link` element and `style` content attribute in a follow up?

Yes, will fix in follow-up
